keps/sig-node/5067-pod-generation/README.md

```diff
@@ -0,0 +1,931 @@
+
+#### Generation and ObservedGeneration
+
+The `metadata.generation` field is currently unused on pods, but we can start 
+leveraging it to help track which pod state is currently being reflected in the 
+pod status. [Generation](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#metadata) is documented as "a sequence 
+number representing a specific generation of the desired state. Set by the 
+system and monotonically increasing, per-resource." For consistency, pod 
+Generation will be incremented whenever the pod has changes that the kubelet 
+needs to actuate. 
+
+A new field ObservedGeneration will be added to the pod status. 
+Kubelet will set this to communicate which pod state is being expressed in the 
+current pod status. This is analogous to the ObservedGeneration that exists
+in other resources' statuses such as [StatefulSets](https://github.com/kubernetes/api/blob/9e7d345b161c12a3056efa88e285b3ef68450c54/apps/v1/types.go#L274).
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+No risks have been identified.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+#### API server
+
+For a newly created pod, the API server will set `generation` to 1. For every 
+update to the pod that requires actuation by the kubelet, the API Server will 
+increment `generation` by 1. Pod updates that result in `generation` being 
+incremented are any changes to mutable fields in the [PodSpec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#podspec-v1-core), and also changes to annotations
+that end up getting copied to the PodSpec such as seccomp and AppArmor annotations.
+
+Pod updates that would not result in `generation` being incremented include:
+* Changes to metadata. This means that if a Pod uses the downward API to make
+pod metadata available to containers, Pod behavior can change without the
+generation being incremented. We will consider this working as intended. 
+* Changes to status.
+
+```
+<<[UNRESOLVED]>>
+Should changes to `DeletionTimestamp` cause `generation` to be incremented? 
```

Comment:
no other objects make deletionTimestamp inform generation, so I would say no

the norm is to compare spec only, and a *few* special-case labels or annotations (I think the annotations that correspond to spec fields like apparmor and seccomp make sense for pods)

Reply:
actually, hmm... there appears to be generic code in place to bump generation on graceful delete:

https://github.com/kubernetes/kubernetes/blob/a444a5bfedb2807632447bfb8350bd41fd1a3d77/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L1013-L1018

https://github.com/kubernetes/kubernetes/blob/a444a5bfedb2807632447bfb8350bd41fd1a3d77/staging/src/k8s.io/apiserver/pkg/registry/rest/delete.go#L153-L159

If this is already done generically, we may not need to do anything special here

Reply:
yeah I just tried this out on a deployment (added a finalizer to a deployment and then issued a `delete`) and it did bump `generation`, so thinking we should just expect the same behavior for pods

cc @tallclair

Reply:
does that actually work today already (does adding a finalizer to a pod and deleting bump the generation?)

Reply:
Oh sorry I missed your question about seeing what happens on graceful delete of a pod. It does actually bump the pod generation today so we can keep that behavior

Reply:
Is the finalizer required to trigger the generation bump? IMO that shouldn't be a requirement for pods. The Kubelet is basically an implicit finalizer, since it needs to actuate the termination of the pod.

Reply:
good question, I'd also expect graceful deletion to set deletionTimestamp and bump the generation

Reply:
Just confirmed experimentally that issuing the delete without adding a finalizer still triggers the generation bump, so no the finalizer is not required